// @secretlint/formatter/lib/formatters/table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------
const chalk_1 = __importDefault(require("chalk"));
const table_1 = require("table");
const strip_ansi_1 = __importDefault(require("strip-ansi"));
const pluralize_1 = __importDefault(require("pluralize"));
//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------
/**
 * Draws text table.
 * @param {Array<Object>} messages Error messages relating to a specific file.
 * @returns {string} A text table.
 */
function drawTable(messages) {
    let rows = [];
    if (messages.length === 0) {
        return "";
    }
    rows.push([
        chalk_1.default.bold("Line"),
        chalk_1.default.bold("Column"),
        chalk_1.default.bold("Type"),
        chalk_1.default.bold("Message"),
        chalk_1.default.bold("Rule ID"),
    ]);
    messages.forEach(function (message) {
        let messageType;
        if (message.severity === "error") {
            messageType = chalk_1.default.red("error");
        }
        else {
            messageType = chalk_1.default.yellow("warning");
        }
        rows.push([message.loc.start.line, message.loc.start.column, messageType, message.message, message.ruleId]);
    });
    const output = table_1.table(rows, {
        columns: {
            0: {
                width: 8,
                wrapWord: true,
            },
            1: {
                width: 8,
                wrapWord: true,
            },
            2: {
                width: 8,
                wrapWord: true,
            },
            3: {
                paddingRight: 5,
                width: 50,
                wrapWord: true,
            },
            4: {
                width: 20,
                wrapWord: true,
            },
        },
        drawHorizontalLine: function (index) {
            return index === 1;
        },
    });
    return output;
}
/**
 * Draws a report (multiple tables).
 * @param {Array} results Report results for every file.
 * @returns {string} A column of text tables.
 */
function drawReport(results) {
    let files;
    files = results.map(function (result) {
        if (!result.messages.length) {
            return "";
        }
        return "\n" + result.filePath + "\n\n" + drawTable(result.messages);
    });
    files = files.filter(function (content) {
        return content.trim();
    });
    return files.join("");
}
//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------
function formatter(report, options) {
    var _a;
    const useColor = (_a = options.color) !== null && _a !== void 0 ? _a : true;
    let output = "";
    let errorCount = 0;
    let warningCount = 0;
    report.forEach((fileReport) => {
        fileReport.messages.forEach((message) => {
            if (message.severity === "warning") {
                warningCount += 1;
            }
            else if (message.severity === "error") {
                errorCount += 1;
            }
        });
    });
    if (errorCount || warningCount) {
        output = drawReport(report);
    }
    output +=
        "\n" +
            table_1.table([
                [chalk_1.default.red(pluralize_1.default("Error", errorCount, true))],
                [chalk_1.default.yellow(pluralize_1.default("Warning", warningCount, true))],
            ], {
                columns: {
                    0: {
                        width: 110,
                        wrapWord: true,
                    },
                },
                drawHorizontalLine: function () {
                    return true;
                },
            });
    if (!useColor) {
        return strip_ansi_1.default(output);
    }
    return output;
}
exports.default = formatter;
