// @secretlint/secretlint-rule-slack/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.creator = exports.messages = void 0;
const regexp_string_matcher_1 = require("@textlint/regexp-string-matcher");
exports.messages = {
    SLACK_TOKEN: {
        en: (props) => `found slack token: ${props.TOKEN}`,
        ja: (props) => `Slackトークン: ${props.TOKEN} がみつかりました`,
    },
    IncomingWebhook: {
        en: (props) => `found Slack Incoming Webhook: ${props.URL}`,
        ja: (props) => `SlackのIncoming Webhooks: ${props.TOKEN} がみつかりました`,
    },
};
function reportIfFoundRawPrivateKey({ source, options, context, t, }) {
    // Based on https://docs.cribl.io/docs/regexesyml
    // https://api.slack.com/docs/token-types
    // Bot user token strings begin with xoxb-
    // User token strings begin with xoxp-
    // Workspace access token strings begin with xoxa-2
    // Workspace refresh token strings begin with xoxr.
    const PRIVATE_KEY_PATTERN = /\bxox[abposr]-(?:\d{1,40}-)+[a-zA-Z0-9]{1,40}\b/g;
    const results = source.content.matchAll(PRIVATE_KEY_PATTERN);
    for (const result of results) {
        const index = result.index || 0;
        const match = result[0] || "";
        const range = [index, index + match.length];
        const allowedResults = regexp_string_matcher_1.matchPatterns(match, options.allows);
        if (allowedResults.length > 0) {
            continue;
        }
        context.report({
            message: t("SLACK_TOKEN", {
                TOKEN: match,
            }),
            range,
        });
    }
}
/**
 * Report if found Incoming Webhooks
 * https://api.slack.com/messaging/webhooks
 * @param source
 * @param options
 * @param context
 * @param t
 */
function reportIfFoundIncomingWebhook({ source, options, context, t, }) {
    // Based on https://hooks.slack.com/TXXXXX/BXXXXX/XXXXXXXXXX
    // https://api.slack.com/messaging/webhooks
    const IncomingWebhooksPattern = /https:\/\/hooks\.slack\.com\/services\/T[a-zA-Z0-9]{1,40}\/B[a-zA-Z0-9]{1,40}\/[a-zA-Z0-9]{1,40}/gi;
    const results = source.content.matchAll(IncomingWebhooksPattern);
    for (const result of results) {
        const index = result.index || 0;
        const match = result[0] || "";
        const range = [index, index + match.length];
        const allowedResults = regexp_string_matcher_1.matchPatterns(match, options.allows);
        if (allowedResults.length > 0) {
            continue;
        }
        context.report({
            message: t("IncomingWebhook", {
                URL: match,
            }),
            range,
        });
    }
}
exports.creator = {
    messages: exports.messages,
    meta: {
        id: "@secretlint/secretlint-rule-slack",
        recommended: true,
        type: "scanner",
        supportedContentTypes: ["text"],
        docs: {
            url: "https://github.com/secretlint/secretlint/blob/master/packages/%40secretlint/secretlint-rule-slack/README.md",
        },
    },
    create(context, options) {
        const t = context.createTranslator(exports.messages);
        const normalizedOptions = {
            allows: options.allows || [],
        };
        return {
            file(source) {
                reportIfFoundRawPrivateKey({ source, options: normalizedOptions, context, t });
                reportIfFoundIncomingWebhook({ source, options: normalizedOptions, context, t });
            },
        };
    },
};
exports.default = exports.creator;
