// @secretlint/secretlint-rule-gcp/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.creator = exports.messages = void 0;
const fs_1 = __importDefault(require("fs"));
const node_forge_1 = __importDefault(require("node-forge"));
const path_1 = __importDefault(require("path"));
exports.messages = {
    PrivateKeyP12: {
        en: (props) => `found GCP Service Account's private key(p12): ${props.FILE_NAME}`,
        ja: (props) => `GCPサービスアカウントの秘密鍵(p12) ${props.FILE_NAME} がみつかりました`,
    },
    PrivateKeyJSON: {
        en: (props) => `found GCP Service Account's private key(json): ${props.FILE_NAME}`,
        ja: (props) => `GCPサービスアカウントの秘密鍵(json): ${props.FILE_NAME} がみつかりました`,
    },
};
function reportIfFoundPrivateKeyJSONFormat({ source, context, t, }) {
    try {
        const credentialObject = JSON.parse(source.content);
        // Private Key Pattern
        const PRIVATE_KEY_PATTERN = /-----BEGIN\s*(DSA|RSA|EC|PGP|OPENSSH)?\s*PRIVATE KEY/gm;
        const isGCPServiceAccountPrivateKeyJSON = "private_key_id" in credentialObject &&
            "private_key" in credentialObject &&
            PRIVATE_KEY_PATTERN.test(credentialObject["private_key"]);
        if (!isGCPServiceAccountPrivateKeyJSON) {
            return;
        }
        context.report({
            message: t("PrivateKeyJSON", {
                FILE_NAME: source.filePath ? path_1.default.basename(source.filePath) : "",
            }),
            range: [0, source.content.length],
        });
    }
    catch (_a) {
        // nope
    }
}
function reportIfFoundPrivateKeyP12Format({ source, context, t, }) {
    if (!source.filePath) {
        return;
    }
    try {
        // Read file as Buffer to Base64 -> bytes -> asn1
        const p12String = fs_1.default.readFileSync(source.filePath).toString("base64");
        const p12Der = node_forge_1.default.util.decode64(p12String);
        const p12Asn1 = node_forge_1.default.asn1.fromDer(p12Der);
        // read p12 file with "notasecret" pass phase
        // The password for Service Account's the PKCS12 file is "notasecret".
        // If success read p12 file, report it as error
        // https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts.keys#serviceaccountprivatekeytype
        node_forge_1.default.pkcs12.pkcs12FromAsn1(p12Asn1, "notasecret");
        // because, this p12 file is credential for GCP Service Account
        context.report({
            message: t("PrivateKeyP12", {
                FILE_NAME: source.filePath ? path_1.default.basename(source.filePath) : "",
            }),
            range: [0, source.content.length],
        });
    }
    catch (_a) {
        // nope
    }
}
exports.creator = {
    messages: exports.messages,
    meta: {
        id: "@secretlint/secretlint-rule-gcp",
        recommended: true,
        type: "scanner",
        supportedContentTypes: ["all"],
        docs: {
            url: "https://github.com/secretlint/secretlint/blob/master/packages/%40secretlint/secretlint-rule-gcp/README.md",
        },
    },
    create(context, options) {
        const t = context.createTranslator(exports.messages);
        const normalizedOptions = {
            allows: options.allows || [],
        };
        return {
            file(source) {
                if (source.ext === ".p12") {
                    reportIfFoundPrivateKeyP12Format({ source, options: normalizedOptions, context, t });
                }
                else if (source.ext === ".json") {
                    reportIfFoundPrivateKeyJSONFormat({ source, options: normalizedOptions, context, t });
                }
            },
        };
    },
};
exports.default = exports.creator;
