// @secretlint/core/lib/RulePresetContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRulePresetContext = void 0;
const RuleContext_1 = require("./RuleContext");
/**
 * Create a context for Rule Preset
 */
const createRulePresetContext = ({ descriptorRulePreset, sourceCode, runningEvents, contextEvents, sharedOptions, locale, }) => {
    const presetRules = descriptorRulePreset.rules || [];
    if (!Array.isArray(presetRules)) {
        console.error(`${descriptorRulePreset.id}:PresetRules is invalid format`, presetRules);
        throw new Error("preset's rules should be an array of rule definitions");
    }
    return {
        sharedOptions,
        registerRule(rule, defaultValue) {
            const descriptorRule = presetRules.find((descriptorRule) => {
                return descriptorRule.id === rule.meta.id;
            });
            // Use undefined instead of {}
            // Default value will be handled by RunningEvents#registerRule
            const descriptorRuleOptions = descriptorRule ? descriptorRule.options : undefined;
            const descriptorRuleDisabled = descriptorRule ? descriptorRule.disabled : undefined;
            const descriptorRuleSeverity = descriptorRule ? descriptorRule.severity : undefined;
            const descriptorAllowMessageIds = descriptorRule ? descriptorRule.allowMessageIds : undefined;
            const context = RuleContext_1.createRuleContext({
                // rule id is a single rule in a preset
                ruleId: rule.meta.id,
                // parent rule id is rule preset id
                ruleParentId: descriptorRulePreset.id,
                meta: rule.meta,
                severity: descriptorRuleSeverity,
                sourceCode,
                contextEvents: contextEvents,
                sharedOptions: sharedOptions,
                locale,
            });
            const defaultValueOfPreset = defaultValue ? defaultValue : {};
            runningEvents.registerRule({
                descriptorRule: Object.assign(Object.assign(Object.assign({}, defaultValueOfPreset), descriptorRule), { id: rule.meta.id, options: descriptorRuleOptions, rule, disabled: descriptorRuleDisabled, allowMessageIds: descriptorAllowMessageIds, severity: descriptorRuleSeverity }),
                context,
            });
        },
    };
};
exports.createRulePresetContext = createRulePresetContext;
